--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3003,6 +_,18 @@
         return (this.getHandle().requestedViewDistance() == 0) ? Bukkit.getViewDistance() : this.getHandle().requestedViewDistance();
     }
 
+    // Sakura start - entity tracking range modifier
+    @Override
+    public double getTrackingRangeModifier() {
+        return this.getHandle().trackingRangeModifier * 100.0;
+    }
+
+    @Override
+    public void setTrackingRangeModifier(double mod) {
+        this.getHandle().trackingRangeModifier = mod / 100.0;
+    }
+    // Sakura end - entity tracking range modifier
+
     // Paper start
     @Override
     public java.util.Locale locale() {
