From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: g0dzcsgo <xdfelgen@gmail.com>
Date: Wed, 20 May 2026 12:10:21 +0200
Subject: [PATCH] Implement native item stacking


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 936f8388c84b3d83a85390ec337a5d0a317838f4..1b3593a55635295497e51408bab46a510a51bc7b 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -396,4 +396,15 @@ public class GlobalConfiguration extends ConfigurationPart {
     public class UpdateChecker extends ConfigurationPart {
         public boolean enabled = true;
     }
+    // Citrus start - Native item stacking
+    public ItemStacking itemStacking;
+        public class ItemStacking extends ConfigurationPart {
+        public boolean enabled = false;
+        @Constraints.Min(1)
+        public int maxStackSize = 1024;
+        @Constraints.Min(0)
+        public double mergeRadius = 2.5D;
+        public boolean resetDespawnTimerOnMerge = true;
+    }
+    // Citrus end - Native item stacking
 }
