From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nahuel Dolores <contact@nahu.me>
Date: Mon, 29 Jun 2026 15:20:54 +0200
Subject: [PATCH] Allow unsigned chat messages


diff --git a/net/minecraft/network/chat/SignedMessageChain.java b/net/minecraft/network/chat/SignedMessageChain.java
index 9395c6a58ba37fb00c492065b8045f54de0a2246..bdbc15d949ae764f99e79530eec07d0e14118e6f 100644
--- a/net/minecraft/network/chat/SignedMessageChain.java
+++ b/net/minecraft/network/chat/SignedMessageChain.java
@@ -57,8 +57,11 @@ public class SignedMessageChain {
                 SignedMessageChain.this.lastTimeStamp = body.timeStamp();
                 PlayerChatMessage unpacked = new PlayerChatMessage(link, signature, body, null, FilterMask.PASS_THROUGH);
                 if (!unpacked.verify(signatureValidator)) {
-                    this.setChainBroken();
-                    throw new SignedMessageChain.DecodeException(SignedMessageChain.DecodeException.INVALID_SIGNATURE);
+                    // Citrus start - Allow unsigned chat messages
+                    //this.setChainBroken();
+                    //throw new SignedMessageChain.DecodeException(SignedMessageChain.DecodeException.INVALID_SIGNATURE);
+                    return unpacked;
+                    // Citrus start - Allow unsigned chat messages
                 }
 
                 if (unpacked.hasExpiredServer(Instant.now())) {
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 61b08dcaca579763d947d289438041554acc5f1e..692624d2c24369a88fab98fb1a34bcd058515359 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -300,11 +300,22 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         this.connection.flushChannel();
     }
 
-    public void send(final Packet<?> packet) {
+    // Citrus start - Allow unsigned chat messages
+    public void send(Packet<?> packet) {
+        if (packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerChatPacket chat) {
+            packet = new net.minecraft.network.protocol.game.ClientboundSystemChatPacket(chat.chatType().decorate(chat.unsignedContent() != null ? chat.unsignedContent() : Component.literal(chat.body().content())), false);
+        }
+    // Citrus end - Allow unsigned chat messages
         this.send(packet, null);
     }
 
     public void send(final Packet<?> packet, final @Nullable ChannelFutureListener listener) {
+        // Citrus start - Allow unsigned chat messages
+        if (packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerChatPacket chat && listener != null) {
+            this.send(chat);
+            return;
+        }
+        // Citrus end - Allow unsigned chat messages
         // CraftBukkit start
         if (packet == null) {
             return;
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index e954bc9d6cd08029c73dfb42804de9570253d8c2..082d431352e71ea3830766d784163127d9fdf2c8 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2805,9 +2805,11 @@ public class ServerGamePacketListenerImpl
                 trackedCount = this.lastSeenMessages.trackedMessagesCount();
             }
 
-            if (trackedCount > 4096) {
-                this.disconnectAsync(Component.translatable("multiplayer.disconnect.too_many_pending_chats"), org.bukkit.event.player.PlayerKickEvent.Cause.TOO_MANY_PENDING_CHATS); // Paper - kick event cause & add proper async disconnect
-            }
+            // Citrus start - Allow unsigned chat messages
+            //if (trackedCount > 4096) {
+            //    this.disconnectAsync(Component.translatable("multiplayer.disconnect.too_many_pending_chats"), org.bukkit.event.player.PlayerKickEvent.Cause.TOO_MANY_PENDING_CHATS); // Paper - kick event cause & add proper async disconnect
+            //}
+            // Citrus end - Allow unsigned chat messages
         }
         }
         // Paper end - Ensure that client receives chat packets in the same order that we add into the message signature cache
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 5bea5dcdaf8ee0a38cd8587a55845e940d8f4cac..4b5bce6f99f5d1ebaccb10498048cda7da97924e 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1155,7 +1155,7 @@ public abstract class PlayerList {
     }
 
     public boolean verifyChatTrusted(final PlayerChatMessage message) {
-        return message.hasSignature() && !message.hasExpiredServer(Instant.now());
+        return true; // message.hasSignature() && !message.hasExpiredServer(Instant.now()); // Citrus - Allow unsigned chat messages
     }
 
     // CraftBukkit start
