From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nahuel Dolores <contact@nahu.me>
Date: Mon, 29 Jun 2026 15:23:49 +0200
Subject: [PATCH] Allow disabled tripwire redstone emissions


diff --git a/net/minecraft/world/level/block/TripWireBlock.java b/net/minecraft/world/level/block/TripWireBlock.java
index 49510dc0ef82621953be8e485708872bd8b2b88b..d97b958859d47d3a531356e3d4c6f9e5a8337977 100644
--- a/net/minecraft/world/level/block/TripWireBlock.java
+++ b/net/minecraft/world/level/block/TripWireBlock.java
@@ -127,7 +127,7 @@ public class TripWireBlock extends Block {
     }
 
     private void updateSource(final Level level, final BlockPos pos, final BlockState state) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent adjacent tripwires from updating
+        // if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent adjacent tripwires from updating // Citrus
         for (Direction direction : new Direction[]{Direction.SOUTH, Direction.WEST}) {
             for (int i = 1; i < 42; i++) {
                 BlockPos testPos = pos.relative(direction, i);
@@ -160,7 +160,7 @@ public class TripWireBlock extends Block {
         final InsideBlockEffectApplier effectApplier,
         final boolean isPrecise
     ) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwires from detecting collision
+        // if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwires from detecting collision // Citrus
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (!level.isClientSide()) {
             if (!state.getValue(POWERED) && !level.getBlockTicks().hasScheduledTick(pos, this)) {
@@ -171,7 +171,7 @@ public class TripWireBlock extends Block {
 
     @Override
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwire pressed check
+        // if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwire pressed check // Citrus
         if (level.getBlockState(pos).getValue(POWERED)) {
             this.checkPressed(level, pos);
         }
@@ -231,7 +231,7 @@ public class TripWireBlock extends Block {
 
         if (shouldBePressed != wasPressed) {
             state = state.setValue(POWERED, shouldBePressed);
-            level.setBlock(pos, state, Block.UPDATE_ALL);
+            // level.setBlock(pos, state, Block.UPDATE_ALL); // Citrus
             this.updateSource(level, pos, state);
         }
 
diff --git a/net/minecraft/world/level/block/TripWireHookBlock.java b/net/minecraft/world/level/block/TripWireHookBlock.java
index 9d08ce61f964bb3115c708091549a233dabcf0e3..07a9a95a08c72e55f42f16b7905fe28af27aea4f 100644
--- a/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -201,7 +201,7 @@ public class TripWireHookBlock extends Block {
                     BlockState wireData = wireStates[i];
                     if (wireData != null) {
                         BlockState testPosState = level.getBlockState(testPos);
-                        if (testPosState.is(Blocks.TRIPWIRE) || testPosState.is(Blocks.TRIPWIRE_HOOK)) {
+                        if (/* testPosState.is(Blocks.TRIPWIRE) ||*/ testPosState.is(Blocks.TRIPWIRE_HOOK)) { // Citrus
                             if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates || !testPosState.is(Blocks.TRIPWIRE)) level.setBlock(testPos, wireData.trySetValue(ATTACHED, attached), Block.UPDATE_ALL); // Paper - prevent tripwire from updating
                         }
                     }
