From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nahuel Dolores <contact@nahu.me>
Date: Mon, 26 Jan 2026 15:59:23 +0100
Subject: [PATCH] Allow disabled tripwire redstone emissions


diff --git a/net/minecraft/world/level/block/TripWireBlock.java b/net/minecraft/world/level/block/TripWireBlock.java
index 9d005c3db8f0c784cdd7217e7b110daa703a9cf3..e95baeca9ecd180ad09cb6deb92c75e53d1ff465 100644
--- a/net/minecraft/world/level/block/TripWireBlock.java
+++ b/net/minecraft/world/level/block/TripWireBlock.java
@@ -128,7 +128,7 @@ public class TripWireBlock extends Block {
     }
 
     private void updateSource(Level level, BlockPos pos, BlockState state) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent adjacent tripwires from updating
+        // if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent adjacent tripwires from updating // Citrus
         for (Direction direction : new Direction[]{Direction.SOUTH, Direction.WEST}) {
             for (int i = 1; i < 42; i++) {
                 BlockPos blockPos = pos.relative(direction, i);
@@ -154,7 +154,7 @@ public class TripWireBlock extends Block {
 
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier, boolean pastEdges) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwires from detecting collision
+        // if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwires from detecting collision // Citrus
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (!level.isClientSide()) {
             if (!state.getValue(POWERED)) {
@@ -165,7 +165,7 @@ public class TripWireBlock extends Block {
 
     @Override
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwire pressed check
+        // if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwire pressed check // Citrus
         if (level.getBlockState(pos).getValue(POWERED)) {
             this.checkPressed(level, pos);
         }
@@ -225,7 +225,7 @@ public class TripWireBlock extends Block {
 
         if (flag != poweredValue) {
             blockState = blockState.setValue(POWERED, flag);
-            level.setBlock(pos, blockState, Block.UPDATE_ALL);
+            // level.setBlock(pos, blockState, Block.UPDATE_ALL); // Citrus
             this.updateSource(level, pos, blockState);
         }
 
diff --git a/net/minecraft/world/level/block/TripWireHookBlock.java b/net/minecraft/world/level/block/TripWireHookBlock.java
index 28869577e480318b5ef469a0b4180b25ddfb627f..a51f2dc4b972dbd8a8e1fe3f282c8c14902e1c6c 100644
--- a/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -205,7 +205,7 @@ public class TripWireHookBlock extends Block {
                     BlockState blockState2 = blockStates[i2];
                     if (blockState2 != null) {
                         BlockState blockState3 = level.getBlockState(blockPos1);
-                        if (blockState3.is(Blocks.TRIPWIRE) || blockState3.is(Blocks.TRIPWIRE_HOOK)) {
+                        if (/*blockState3.is(Blocks.TRIPWIRE) ||*/ blockState3.is(Blocks.TRIPWIRE_HOOK)) { // Citrus
                             if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates || !blockState3.is(Blocks.TRIPWIRE))  level.setBlock(blockPos1, blockState2.trySetValue(ATTACHED, flag2), Block.UPDATE_ALL); // Paper - prevent tripwire from updating
                         }
                     }
