From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nahuel Dolores <contact@nahu.me>
Date: Mon, 11 May 2026 16:29:15 +0200
Subject: [PATCH] Allow unsigned chat messages


diff --git a/net/minecraft/network/chat/SignedMessageChain.java b/net/minecraft/network/chat/SignedMessageChain.java
index 99a8a8b5cb65bd94d9496ddb6ff11b8a67f3a870..e57187e048dc6503414cac95cd6bd6df7bdae127 100644
--- a/net/minecraft/network/chat/SignedMessageChain.java
+++ b/net/minecraft/network/chat/SignedMessageChain.java
@@ -56,8 +56,11 @@ public class SignedMessageChain {
                         SignedMessageChain.this.lastTimeStamp = body.timeStamp();
                         PlayerChatMessage unpacked = new PlayerChatMessage(link, signature, body, null, FilterMask.PASS_THROUGH);
                         if (!unpacked.verify(signatureValidator)) {
-                            this.setChainBroken();
-                            throw new SignedMessageChain.DecodeException(SignedMessageChain.DecodeException.INVALID_SIGNATURE);
+                            // Citrus start - Allow unsigned chat messages
+                            //this.setChainBroken();
+                            //throw new SignedMessageChain.DecodeException(SignedMessageChain.DecodeException.INVALID_SIGNATURE);
+                            return unpacked;
+                            // Citrus end - Allow unsigned chat messages
                         } else {
                             if (unpacked.hasExpiredServer(Instant.now())) {
                                 SignedMessageChain.LOGGER.warn("Received expired chat: '{}'. Is the client/server system time unsynchronized?", body.content());
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index f816ceb53ab72c22cabe782f3f20af43733090e1..c4a9e9f1aa18a16b6dd3b1f1635f4a3b21f6f67f 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -300,11 +300,22 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         this.connection.flushChannel();
     }
 
-    public void send(final Packet<?> packet) {
+    // Citrus start - Allow unsigned chat messages
+    public void send(Packet<?> packet) {
+        if (packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerChatPacket chat) {
+            packet = new net.minecraft.network.protocol.game.ClientboundSystemChatPacket(chat.chatType().decorate(chat.unsignedContent() != null ? chat.unsignedContent() : Component.literal(chat.body().content())), false);
+        }
+    // Citrus end - Allow unsigned chat messages
         this.send(packet, null);
     }
 
     public void send(final Packet<?> packet, final @Nullable ChannelFutureListener listener) {
+        // Citrus start - Allow unsigned chat messages
+        if (packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerChatPacket chat && listener != null) {
+            this.send(chat);
+            return;
+        }
+        // Citrus end - Allow unsigned chat messages
         // CraftBukkit start
         if (packet == null) {
             return;
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 1b2a079ffc7c0c9290b471768dfccf61e5ba996b..2730db6b2788b970283addc4053ef34d9953355e 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2803,9 +2803,11 @@ public class ServerGamePacketListenerImpl
                 trackedCount = this.lastSeenMessages.trackedMessagesCount();
             }
 
-            if (trackedCount > 4096) {
-                this.disconnectAsync(Component.translatable("multiplayer.disconnect.too_many_pending_chats"), org.bukkit.event.player.PlayerKickEvent.Cause.TOO_MANY_PENDING_CHATS); // Paper - kick event cause & add proper async disconnect
-            }
+            // Citrus start - Allow unsigned chat messages
+            //if (trackedCount > 4096) {
+            //    this.disconnectAsync(Component.translatable("multiplayer.disconnect.too_many_pending_chats"), org.bukkit.event.player.PlayerKickEvent.Cause.TOO_MANY_PENDING_CHATS); // Paper - kick event cause & add proper async disconnect
+            //}
+            // Citrus end - Allow unsigned chat messages
         }
         }
         // Paper end - Ensure that client receives chat packets in the same order that we add into the message signature cache
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 82748af2479b0f8f15e39d1ad1ccd2a0e583bcdd..c89edf35d59e4bd423165745a2126a2fa6557866 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1158,7 +1158,7 @@ public abstract class PlayerList {
     }
 
     public boolean verifyChatTrusted(final PlayerChatMessage message) {
-        return message.hasSignature() && !message.hasExpiredServer(Instant.now());
+        return true; // Citrus - Allow unsigned chat messages
     }
 
     // CraftBukkit start
