From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Fri, 8 Dec 2023 18:21:56 +0000
Subject: [PATCH] Add instant mob death animation


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index acfb7762f53636a5df1e8de8f1cbcf8d150debc7..dc57d8d4448f60478693b6c3e81ae11074d4f22b 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1858,6 +1858,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
             // Paper start
             if (this.dead) { // Paper
+                // Sakura start
+                if (level().sakuraConfig().entity.instantDeathAnimation && !(this instanceof Player)) {
+                    this.deathTime = 20;
+                    return;
+                }
+                // Sakura end
                 this.level().broadcastEntityEvent(this, (byte) 3);
 
             this.setPose(Pose.DYING);
