From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Tue, 20 Feb 2024 19:16:16 +0000
Subject: [PATCH] Add entity travel distance limits


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index a9315006379f7ab5b40cb6aba7ed82727afe3a5a..292ec3f8332631c6614e5ed25eeb82f9c2615a3b 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1306,6 +1306,11 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         final boolean isActive = org.spigotmc.ActivationRange.checkIfActive(entity); // Paper - EAR 2
         if (isActive) { // Paper - EAR 2
         entity.tick();
+            // Sakura start - entity travel distance limits
+            if (entity.isPastTravelDistanceLimit()) {
+                entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+            }
+            // Sakura end - entity travel distance limits
         entity.postTick(); // CraftBukkit
         } else { entity.inactiveTick(); } // Paper - EAR 2
         gameprofilerfiller.pop();
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 80451700fac1da06401097c80d416100dd331c31..17255cc9739392d23b5b55f2be5660fc56d538d5 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -680,6 +680,19 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return this.physics;
     }
     // Sakura end - physics version api
+    // Sakura start - entity travel distance limits
+    private final double travelDistanceLimit;
+
+    public final boolean isPastTravelDistanceLimit() {
+        if (this.origin == null) {
+            return false;
+        }
+
+        double x = Math.pow(this.origin.getX() - this.position.x(), 2);
+        double z = Math.pow(this.origin.getZ() - this.position.z(), 2);
+        return Math.max(x, z) >= this.travelDistanceLimit;
+    }
+    // Sakura end - entity travel distance limits
 
     public Entity(EntityType<?> type, Level world) {
         this.id = Entity.ENTITY_COUNTER.incrementAndGet();
@@ -732,6 +745,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         this.entityData = datawatcher_a.build();
         this.setPos(0.0D, 0.0D, 0.0D);
         this.eyeHeight = this.dimensions.eyeHeight();
+        this.travelDistanceLimit = Math.pow(this.level.sakuraConfig().entity.chunkTravelLimit.getOrDefault(type, Integer.MAX_VALUE) * 16.0, 2); // Sakura - entity travel distance limits
     }
 
     public boolean isColliding(BlockPos pos, BlockState state) {
