From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 23 Feb 2024 02:07:03 +0000
Subject: [PATCH] Change shields to reduce damage


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 365ab6a0c564911bc58d3fb90e3abe82dbbdc642..8992f0ad55c1177bee9d4167cedf2fef54117a6a 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2433,7 +2433,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
         com.google.common.base.Function<Double, Double> blocking = new com.google.common.base.Function<Double, Double>() {
             @Override
             public Double apply(Double f) {
+                    // Sakura start - shield damage reduction
+                    if (!level().sakuraConfig().players.combat.shieldDamageReduction || damagesource.getDirectEntity() instanceof AbstractArrow) {
                 return -((LivingEntity.this.isDamageSourceBlocked(damagesource)) ? f : 0.0);
+                    } else {
+                        return -(LivingEntity.this.isBlocking() ? f * 0.5 : 0.0);
+                    }
+                    // Sakura end - shield damage reduction
             }
         };
         float blockingModifier = blocking.apply((double) f).floatValue();
