From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 23 Feb 2024 02:07:03 +0000
Subject: [PATCH] Change shields to reduce damage


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 3534821c8b600b9b3d3655c071fc59a3e3b4c23e..88dc63e5c739eb39822999d226856205a3c6099b 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2416,7 +2416,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
         com.google.common.base.Function<Double, Double> blocking = new com.google.common.base.Function<Double, Double>() {
             @Override
             public Double apply(Double f) {
+                    // Sakura start - shield damage reduction
+                    if (!level().sakuraConfig().players.combat.shieldDamageReduction || damagesource.getDirectEntity() instanceof AbstractArrow) {
                 return -((LivingEntity.this.isDamageSourceBlocked(damagesource)) ? f : 0.0);
+                    } else {
+                        return -(LivingEntity.this.isBlocking() ? f * 0.5 : 0.0);
+                    }
+                    // Sakura end - shield damage reduction
             }
         };
         float blockingModifier = blocking.apply((double) f).floatValue();
