From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Mon, 29 Jul 2024 00:18:58 +0100
Subject: [PATCH] Set entity impulse on explosion


diff --git a/src/main/java/me/samsuik/sakura/explosion/special/SpecialisedExplosion.java b/src/main/java/me/samsuik/sakura/explosion/special/SpecialisedExplosion.java
index 449fa5518a1f9013dbd41dc4bfe3ff8a57b94a30..e4f4ae1ce4703e93664e3c2b97a8cae1e62aaebd 100644
--- a/src/main/java/me/samsuik/sakura/explosion/special/SpecialisedExplosion.java
+++ b/src/main/java/me/samsuik/sakura/explosion/special/SpecialisedExplosion.java
@@ -203,6 +203,7 @@ public abstract class SpecialisedExplosion<T extends Entity> extends ServerExplo
             moveZ += z;
         }
 
+        entity.hasImpulse = true; // Sakura - set entity impulse on explosion
         entity.setDeltaMovement(moveX, moveY, moveZ);
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/ServerExplosion.java b/src/main/java/net/minecraft/world/level/ServerExplosion.java
index a82922cd0e648aa1154638a7ff5c616d37475f9c..4109a781b2882e24a20ec5a5ca87040ad0d99bc3 100644
--- a/src/main/java/net/minecraft/world/level/ServerExplosion.java
+++ b/src/main/java/net/minecraft/world/level/ServerExplosion.java
@@ -748,6 +748,7 @@ public class ServerExplosion implements Explosion {
                            // Paper end - knockback events
                         }
                         // CraftBukkit end
+                        entity.hasImpulse = true; // Sakura - set entity impulse on explosion
                         entity.setDeltaMovement(entity.getDeltaMovement().add(vec3d));
                         if (entity instanceof Player) {
                             Player entityhuman = (Player) entity;
