From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Wed, 2 Oct 2024 18:45:22 +0100
Subject: [PATCH] Add max armour durability damage


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 8992f0ad55c1177bee9d4167cedf2fef54117a6a..8cb19ec587512d70a792235ba4e7d82b56074a77 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2535,6 +2535,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
             // Apply damage to armor
             if (!damagesource.is(DamageTypeTags.BYPASSES_ARMOR)) {
                 float armorDamage = (float) (event.getDamage() + event.getDamage(DamageModifier.BLOCKING) + event.getDamage(DamageModifier.HARD_HAT));
+                // Sakura start - add max armour durability damage
+                int maxArmourDamage = this.level().sakuraConfig().players.combat.maxArmourDamage.or(-1);
+                if (maxArmourDamage >= 0) {
+                    armorDamage = Math.min(armorDamage, maxArmourDamage);
+                }
+                // Sakura end - add max armour durability damage
                 this.hurtArmor(damagesource, armorDamage);
             }
 
