From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nahuel Dolores <contact@nahu.me>
Date: Wed, 18 Sep 2024 10:54:50 +0200
Subject: [PATCH] Add support for stackable buckets


diff --git a/src/main/java/me/samsuik/sakura/configuration/GlobalConfiguration.java b/src/main/java/me/samsuik/sakura/configuration/GlobalConfiguration.java
index 5296d4e0a1041932e36562f42fbf3e1dcfea33fe..31f37fdaf658feed820f511c9c2d424ca3788bb9 100644
--- a/src/main/java/me/samsuik/sakura/configuration/GlobalConfiguration.java
+++ b/src/main/java/me/samsuik/sakura/configuration/GlobalConfiguration.java
@@ -11,7 +11,7 @@ import org.spongepowered.configurate.objectmapping.meta.Setting;
 public class GlobalConfiguration extends ConfigurationPart {
 
     private static final Logger LOGGER = LogUtils.getClassLogger();
-    static final int CURRENT_VERSION = 1;// (when you change the version, change the comment, so it conflicts on rebases): rename filter bad nbt from spawn eggs
+    static final int CURRENT_VERSION = 2;// (when you change the version, change the comment, so it conflicts on rebases): buckets !
 
     private static GlobalConfiguration instance;
     public static GlobalConfiguration get() {
@@ -25,6 +25,12 @@ public class GlobalConfiguration extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     public int version = CURRENT_VERSION;
 
+    public StackableBuckets stackableBuckets;
+    public class StackableBuckets extends ConfigurationPart {
+        public boolean enabled = true;
+        public int maxBucketStackSize = 16;
+    }
+
     public Fps fps;
     public class Fps extends ConfigurationPart {
         public String message = "<dark_gray>(<light_purple>S</light_purple>) <gray><state> <yellow><name>";
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 5f066ffa28b6eb6560c4645ce52f7cbc36c7d3f5..556aa40d867406476c242d0f416664fcdd0d7b19 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -3222,6 +3222,18 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
                         switch (event.getResult()) {
                             case ALLOW:
                             case DEFAULT:
+                                // Sakura start - Add support for stackable spawners
+                                if (me.samsuik.sakura.configuration.GlobalConfiguration.get().stackableBuckets.enabled) {
+                                    if (itemstack.getItem() == Items.LAVA_BUCKET || itemstack.getItem() == Items.WATER_BUCKET || itemstack.getItem() == Items.MILK_BUCKET) {
+                                        if (action == InventoryAction.MOVE_TO_OTHER_INVENTORY) {
+                                            this.player.containerMenu.sendAllDataToRemote();
+                                        } else {
+                                            this.player.connection.send(new ClientboundContainerSetSlotPacket(-1, -1, this.player.inventoryMenu.incrementStateId(), this.player.containerMenu.getCarried()));
+                                            this.player.connection.send(new ClientboundContainerSetSlotPacket(this.player.containerMenu.containerId, this.player.inventoryMenu.incrementStateId(), packet.getSlotNum(), this.player.containerMenu.getSlot(packet.getSlotNum()).getItem()));
+                                        }
+                                    }
+                                }
+                                // Sakura end
                                 this.player.containerMenu.clicked(i, packet.getButtonNum(), packet.getClickType(), this.player);
                                 break;
                             case DENY:
diff --git a/src/main/java/net/minecraft/world/item/Items.java b/src/main/java/net/minecraft/world/item/Items.java
index 07315232192f6e09910a028c4643d7f0544c62e3..28225d5af8155008893cbbcafdf8c3f24f121635 100644
--- a/src/main/java/net/minecraft/world/item/Items.java
+++ b/src/main/java/net/minecraft/world/item/Items.java
@@ -1210,14 +1210,14 @@ public class Items {
         "warped_hanging_sign", new HangingSignItem(Blocks.WARPED_HANGING_SIGN, Blocks.WARPED_WALL_HANGING_SIGN, new Item.Properties().stacksTo(16))
     );
     public static final Item BUCKET = registerItem("bucket", new BucketItem(Fluids.EMPTY, new Item.Properties().stacksTo(16)));
-    public static final Item WATER_BUCKET = registerItem("water_bucket", new BucketItem(Fluids.WATER, new Item.Properties().craftRemainder(BUCKET).stacksTo(1)));
-    public static final Item LAVA_BUCKET = registerItem("lava_bucket", new BucketItem(Fluids.LAVA, new Item.Properties().craftRemainder(BUCKET).stacksTo(1)));
+    public static final Item WATER_BUCKET = registerItem("water_bucket", new BucketItem(Fluids.WATER, new Item.Properties().craftRemainder(BUCKET).stacksTo(16))); // Sakura - Add support for stackable buckets
+    public static final Item LAVA_BUCKET = registerItem("lava_bucket", new BucketItem(Fluids.LAVA, new Item.Properties().craftRemainder(BUCKET).stacksTo(16))); // Sakura - Add support for stackable buckets
     public static final Item POWDER_SNOW_BUCKET = registerItem(
         "powder_snow_bucket", new SolidBucketItem(Blocks.POWDER_SNOW, SoundEvents.BUCKET_EMPTY_POWDER_SNOW, new Item.Properties().stacksTo(1))
     );
     public static final Item SNOWBALL = registerItem("snowball", new SnowballItem(new Item.Properties().stacksTo(16)));
     public static final Item LEATHER = registerItem("leather", new Item(new Item.Properties()));
-    public static final Item MILK_BUCKET = registerItem("milk_bucket", new MilkBucketItem(new Item.Properties().craftRemainder(BUCKET).stacksTo(1)));
+    public static final Item MILK_BUCKET = registerItem("milk_bucket", new MilkBucketItem(new Item.Properties().craftRemainder(BUCKET).stacksTo(16))); // Sakura - Add support for stackable buckets
     public static final Item PUFFERFISH_BUCKET = registerItem(
         "pufferfish_bucket",
         new MobBucketItem(
